src/traffic_server/traffic_server.cc

```diff
@@ -223,9 +223,19 @@ struct AutoStopCont : public Continuation {
   mainEvent(int /* event */, Event * /* e */)
   {
     APIHook *hook = lifecycle_hooks->get(TS_LIFECYCLE_SHUTDOWN_HOOK);
-    while (hook) {
-      hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
-      hook = hook->next();
+    if (hook != nullptr) {
+      if (hook->m_cont->mutex != nullptr) {
+        MUTEX_TAKE_LOCK(hook->m_cont->mutex, this_ethread());
+        while (hook) {
+          hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
+          hook = hook->next();
+        }
+      } else {
+        while (hook) {
+          hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
+          hook = hook->next();
+        }
+      }
```

Comment:
I think what you meant is:
```
while (hook) {
  if (hook->m_cont->mutex) {
    SCOPED_MUTEX_LOCK(g, hook->m_cont->mutex, this_ethread());
    hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
  } else {
    hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
  }
  hook = hook->next();
}
```

Reply:
yeah, also just remembered that hook changes.